### DISCUSSION TITLE
Retrieving Config from Recursively Instantiated Objects in Hydra

### DISCUSSION THREAD

**Discussion by juergenmathes** [facebookresearch/hydra#2730] [Q&A]
Hello community,

I'm currently working with the Hydra configuration management library in Python. My use-case involves recursively instantiating several classes with Hydra configs, and at a later point, I need to retrieve the configuration from these instantiated objects, including their nested instances.

What I am looking for is a convenient, possibly built-in, way to retrieve the config associated with an instance and its nested instances, ideally in a format that is compatible with Hydra or `DictConfig`. 

Here's a simple illustration of what I'm doing:

```
@hydra.main(config_path="conf", config_name="config")
def my_app(cfg : DictConfig) -> None:
    my_instance = MyClass(cfg) 
    # ... Some operations that may change the state of my_instance ...
    # At this point, I want to retrieve the config from my_instance and its potential sub-instances.
    retrieved_cfg = my_instance.to_config()
    assert retrieved_cfg == cfg
    # By doing this, I hope to maintain the integrity and consistency of my application's configuration.
    # Also, this would offer me the ability to create a config starting from code, which may also be type checked using mypy. 
    # Users who are not familiar with the code base can just generate their own configs from an already instantiated object and then dump it to a file as a blue print for their experiments.

if __name__ == "__main__":
    my_app()
```
    
    
 What I would like to avoid is writing a to_config method for every class and having to account for recursive object relationships manually, like so:
 
```
 class MyClass:
    def to_config(self) -> DictConfig:
        # Here, I would need to manually retrieve configs for each sub-instance
       ....
```

This seems inefficient and unscalable, especially as the complexity of object relationships grows.

Does Hydra offer a built-in method or a more elegant way to extract the config from an instance, including its sub-instances, in a DictConfig format or any other Hydra-related format?

Any pointers or guidance would be greatly appreciated. Thank you.

**Comment by odelalleau** [comment]
I'm not sure whether Hydra provides a way to do it automatically. Couple of thoughts:

1. You can do something like
```python
def instantiate(cfg):
    obj = hydra.utils.instantiate(cfg)
    obj.get_config = lambda: cfg
    return obj
```
which will let you call `my_obj.get_config()` after the object is instantiated (since you are talking about recursively instantiated objects I assume you are using the recursive instantiation feature from `hydra.utils.instantiate()` rather than the code snipped you provided in your post).

A potential drawback is that you're getting back the config that was used to create the object(s), but if some fields have been updated you won't know about it. That being said, what you could do is use that config to figure out the tree structure and write a function that walks through it to inspect each object within it and reconstruct the "real" config based on the current values stored in those objects.

2. You could look into [hydra-zen](https://mit-ll-responsible-ai.github.io/hydra-zen/index.html#). I'm not familiar with it but I wonder if it might help you with this kind of stuff.

**Comment by juergenmathes** [comment]
Hmm.. thanks. I count this as if hydra has not to purpose to provide such functionality.

**Reply by odelalleau** [reply]
You can always open a feature request issue but the best way to make it happen is probably to implement it yourself ;)
